File: core/api/src/test/java/ru/endlesscode/rpginventory/ConfigurationCollectorTest.java
package ru.endlesscode.rpginventory;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import ru.endlesscode.rpginventory.configuration.ConfigurationCollector;
import ru.endlesscode.rpginventory.item.ConfigurableItemStack;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.*;

public class ConfigurationCollectorTest {

    private final Map<String, String> stringValues = new HashMap<String, String>() {{
        this.put("first", "Nulla semper facilisis urna non fermentum.");
        this.put("second", "Morbi at lorem vitae odio molestie scelerisque.");
        this.put("third", "Vivamus non neque nec purus auctor hendrerit.");
        this.put("fourth", "Integer nec auctor ipsum, porttitor dictum sapien.");
    }};

    private final Map<String, ConfigurableItemStack> cisValues = new HashMap<String, ConfigurableItemStack>() {{
        this.put("magicStick", ConfigurableItemStack.Builder.fromMaterial("STICK")
                .withDisplayName("&6Magic stick").withLore(
                        Arrays.asList(
                                "&7This stick can be obtained in",
                                "&7the &cElite dungeon&7 after defeating",
                                "&7a &4Bloody swordmaster&7."
                        )
                ).build()
        );
        this.put("uselessStick", ConfigurableItemStack.Builder.fromMaterial("STICK")
                .withDisplayName("&7Useless stick").withLore(
                        Arrays.asList(
                                "&7This stick can be obtained everywhere,",
                                "&7where a tree available."
                        )
                ).build()
        );
        this.put("justStick", ConfigurableItemStack.Builder.fromMaterial("STICK")
                .withDisplayName("&aJust stick").withLore(
                        Collections.singletonList("&7Where you found it?..")
                ).build()
        );
    }};

    private Path tmpDir;

    @Before
    public void setUp() throws Exception {
        Path testDir = Files.createDirectories(Paths.get("testFiles"));
        this.tmpDir = Files.createTempDirectory(testDir, "cfg");
    }

    @Test
    public void stringValuesRead() {
        this.saveResource(this.tmpDir, "stringValues.conf");
        final ConfigurationCollector collector = new ConfigurationCollector(this.tmpDir.toFile());
        final Map<String, String> collect = collector.collect(String.class, String.class);
        if (!collect.keySet().containsAll(this.stringValues.keySet())) {
            Assert.fail();
        }
        if (!collect.values().containsAll(this.stringValues.values())) {
            Assert.fail();
        }
    }

    @Test
    public void cisValuesRead() {
        this.saveResource(this.tmpDir, "cisValues.conf");
        final ConfigurationCollector collector = new ConfigurationCollector(this.tmpDir.toFile());
        final Map<String, ConfigurableItemStack> collect = collector.collect(String.class, ConfigurableItemStack.class);
        System.out.println(collect);
        if (!collect.keySet().containsAll(this.stringValues.keySet())) {
            Assert.fail();
        }
        if (!collect.values().containsAll(this.stringValues.values())) {
            Assert.fail();
        }
    }

    @After
    public void tearDown() throws Exception {
        Files.walk(this.tmpDir)
                .sorted(Comparator.reverseOrder())
                .forEach(this::deleteFile);
    }

    private void deleteFile(Path path) {
        try {
            Files.delete(path);
        } catch (IOException ignored) {
        }
    }

    private void saveResource(Path targetDirectory, String name) {
        InputStream resourceAsStream = this.getClass().getClassLoader().getResourceAsStream(name);
        if (resourceAsStream == null) {
            return;
        }
        final Path resolve = targetDirectory.resolve(name);
        try {
            Files.copy(resourceAsStream, resolve, StandardCopyOption.REPLACE_EXISTING);
        } catch (IOException ignore) {
            System.err.println("Failed to save" + name);
        }

    }
}

File: core/src/test/kotlin/internal/config/ConfigurationCollectorTest.kt
/*
 * This file is part of ECInventory
 * <https://github.com/EndlessCodeGroup/ECInventory>.
 * Copyright (c) 2019-2021 EndlessCode Group and contributors
 *
 * ECInventory is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * ECInventory is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with ECInventory. If not, see <http://www.gnu.org/licenses/>.
 */

package ru.endlesscode.inventory.internal.config

import io.kotest.matchers.maps.shouldContainExactly
import io.kotest.matchers.shouldBe
import ru.endlesscode.inventory.internal.data.DataConfig
import ru.endlesscode.inventory.internal.data.InventoryConfig
import ru.endlesscode.inventory.internal.data.SlotConfig
import ru.endlesscode.inventory.internal.util.copyTo
import ru.endlesscode.inventory.slot.Slot
import ru.endlesscode.inventory.test.FileTestBase
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import kotlin.test.Test

class ConfigurationCollectorTest : FileTestBase() {

    // SUT
    private val collector = ConfigurationCollector(dir)

    @Test
    fun `when collect string values - should return right values`() {
        // Given
        saveResource(dir, "config/stringValues.conf")

        // When
        val collected = collector.collect<Map<String, String>>()

        // Then
        collected shouldContainExactly mapOf(
            "first" to "Nulla semper facilisis urna non fermentum.",
            "second" to "Morbi at lorem vitae odio molestie scelerisque.",
            "third" to "Vivamus non neque nec purus auctor hendrerit.",
            "fourth" to "Integer nec auctor ipsum, porttitor dictum sapien.",
        )
    }

    @Test
    fun `when collect DataConfig - should return right values`() {
        // Given
        saveResource(dir, "config/data.conf")

        // When
        val collected = collector.collect<DataConfig>()

        // Then
        collected shouldBe DataConfig(
            slots = mapOf(
                "right-ring" to SlotConfig(
                    name = "Right ring",
                    texture = "ring-slot",
                    allowedItems = listOf("minecraft:diamond_shovel", "mimic:some_texture_item"),
                    type = Slot.Type.PASSIVE,
                    maxStackSize = 1,
                ),
                "left-ring" to SlotConfig(
                    name = "Left ring",
                    texture = "ring-slot",
                    allowedItems = listOf("minecraft:diamond_shovel", "mimic:some_texture_item"),
                    type = Slot.Type.PASSIVE,
                    maxStackSize = 1,
                ),
            ),
            inventories = mapOf(
                "default" to InventoryConfig(
                    name = "ECInventory",
                    emptySlotTexture = null,
                    slots = mapOf(
                        "24" to "left-ring",
                        "26" to "right-ring",
                    ),
                ),
            )
        )
    }

    private fun saveResource(targetDirectory: Path, name: String) {
        val resourceAsStream = javaClass.classLoader.getResourceAsStream(name) ?: return
        val resolve = targetDirectory.resolve(name.substringAfter('/'))
        resourceAsStream.copyTo(resolve, StandardCopyOption.REPLACE_EXISTING)
    }
}
